File: C/Thread/5.11.c
#include <stdio.h>
#include <pthread.h>
#include <stdlib.h>
#include <unistd.h>

#define SIZE 15

int data[SIZE];     // base de données (tableau)
int read_count = 0; // nombre de lecteurs actifs

pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER; // protège read_count
pthread_mutex_t wrt = PTHREAD_MUTEX_INITIALIZER;   // protège la base


void* reader(void* arg)
{
    int id = *(int*)arg;

    while(1)
    {
        // on protège read_count
        pthread_mutex_lock(&mutex);
        read_count++;

        if(read_count == 1)
            pthread_mutex_lock(&wrt); // premier lecteur bloque écrivains

        pthread_mutex_unlock(&mutex);

        // lecture (pas critique)
        int i = rand() % SIZE;
        printf("Lecteur %d lit data[%d] = %d\n", id, i, data[i]);

        sleep(rand()%3 + 1);

        // fin lecture
        pthread_mutex_lock(&mutex);
        read_count--;

        if(read_count == 0)
            pthread_mutex_unlock(&wrt); // dernier lecteur libère écrivains

        pthread_mutex_unlock(&mutex);
    }

    return NULL;
}


void* writer(void* arg)
{
    int id = *(int*)arg;

    while(1)
    {
        pthread_mutex_lock(&wrt); // accès exclusif

        int i = rand() % SIZE;
        data[i] = rand() % 100;

        printf("Écrivain %d écrit data[%d] = %d\n", id, i, data[i]);

        pthread_mutex_unlock(&wrt);

        sleep(rand()%5 + 1);
    }

    return NULL;
}


int main()
{
    srand(time(NULL));

    int N = 3; // lecteurs
    int M = 2; // écrivains

    pthread_t readers[N], writers[M];
    int ids[N > M ? N : M];

    // création lecteurs
    for(int i=0;i<N;i++)
    {
        ids[i] = i;
        pthread_create(&readers[i], NULL, reader, &ids[i]);
    }

    // création écrivains
    for(int i=0;i<M;i++)
    {
        ids[i] = i;
        pthread_create(&writers[i], NULL, writer, &ids[i]);
    }

    // attente
    for(int i=0;i<N;i++)
        pthread_join(readers[i], NULL);

    for(int i=0;i<M;i++)
        pthread_join(writers[i], NULL);

    return 0;
}

File: Python/Thread/5.11.py
import threading
import time
import random


SIZE = 15
data = [0] * SIZE
read_count = 0

mutex = threading.Lock()  # protège read_count
wrt = threading.Lock()    # accès exclusif pour écrivains

def reader(id):
    global read_count

    while True:
        # début section critique
        with mutex:
            read_count += 1
            if read_count == 1:
                wrt.acquire()  # premier lecteur bloque écrivains

        # lecture
        i = random.randint(0, SIZE-1)
        print(f"Lecteur {id} lit data[{i}] = {data[i]}")
        time.sleep(random.randint(1,3))

        # fin lecture
        with mutex:
            read_count -= 1
            if read_count == 0:
                wrt.release()  # dernier lecteur libère écrivains

def writer(id):
    while True:
        wrt.acquire()  # accès exclusif

        i = random.randint(0, SIZE-1)
        data[i] = random.randint(0, 100)
        print(f"Écrivain {id} écrit data[{i}] = {data[i]}")

        wrt.release()
        time.sleep(random.randint(1,5))

N = 3  # lecteurs
M = 2  # écrivains

for i in range(N):
    threading.Thread(target=reader, args=(i,), daemon=True).start()

for i in range(M):
    threading.Thread(target=writer, args=(i,), daemon=True).start()

while True:
    time.sleep(1)